test/Dialect/FIRRTL/lower-layers.mlir

```diff
@@ -1,11 +1,366 @@
+  // CHECK:   firrtl.connect %w, %[[p]] : !firrtl.uint<1>, !firrtl.uint<1>
+  // CHECK: }
+  // CHECK: firrtl.module @CaptureHardwareViaConnect() {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %[[p:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A]]
+  // CHECK:   firrtl.strictconnect %[[p]], %c0_ui1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @CaptureHardwareViaConnect() {
+    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %w = firrtl.wire : !firrtl.uint<1>
+      firrtl.connect %w, %c0_ui1 : !firrtl.uint<1>, !firrtl.uint<1>
+    }
+  }
+
+  // CHECK: firrtl.module private @[[A:.+]](in %[[p:.+]]: !firrtl.uint<1>) {
+  // CHECK:   %0 = firrtl.ref.send %[[p]] : !firrtl.uint<1>
+  // CHECK:   %1 = firrtl.ref.resolve %0 : !firrtl.probe<uint<1>>
+  // CHECK: }
+  // CHECK: firrtl.module @CaptureProbeSrc() {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %w = firrtl.wire : !firrtl.uint<1>
+  // CHECK:   %0 = firrtl.ref.send %w : !firrtl.uint<1>
+  // CHECK:   %[[p:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A]]
+  // CHECK:   %1 = firrtl.ref.resolve %0 : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.strictconnect %[[p]], %1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @CaptureProbeSrc() {
+    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    %w = firrtl.wire : !firrtl.uint<1>
+    %r = firrtl.ref.send %w : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      firrtl.ref.resolve %r : !firrtl.probe<uint<1>>
+    }
+  }
+
+  // CHECK: firrtl.module private @[[B:.+]](in %[[p:.+]]: !firrtl.uint<1>, in %[[q:.+]]: !firrtl.uint<1>)
+  // CHECK:   %0 = firrtl.add %[[p]], %[[q]] : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+  // CHECK: }
+  // CHECK: firrtl.module private @[[A:.+]](out %[[p:.+]]: !firrtl.probe<uint<1>>, out %[[q:.+]]: !firrtl.probe<uint<1>>) attributes {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %0 = firrtl.ref.send %c0_ui1 : !firrtl.uint<1>
+  // CHECK:   firrtl.ref.define %[[q]], %0 : !firrtl.probe<uint<1>>
+  // CHECK:   %c0_ui1_1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %1 = firrtl.ref.send %c0_ui1_1 : !firrtl.uint<1>
+  // CHECK:   firrtl.ref.define %[[p]], %1 : !firrtl.probe<uint<1>>
+  // CHECK: }
+  // CHECK: firrtl.module @NestedCaptureHardware() {
+  // CHECK:   %[[b1:.+]], %[[b2:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A_B.sv", excludeFromFileList>} @[[B]]
+  // CHECK:   %[[a1:.+]], %[[a2:.+]] = firrtl.instance {{.+}} {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @[[A]]
+  // CHECK:   %0 = firrtl.ref.resolve %[[a2]] : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.strictconnect %[[b1]], %0 : !firrtl.uint<1>
```

Comment:
Some of these couplets could be cleaned up by the pass instead of generating them. This is entirely follow-on work!